### PR TITLE
Add default /etc/aliases for amavis, clamav.  Add bsd-mailx.

### PR DESCRIPTION
**Pull request by MakerMatrix**
Also fixed dovecot keygen so it doesn't generate an error msg about RANDFILE.

### CODE REVIEW

**Comment by aendeavor** [comment]
I like this changes - thanks for creating this PR:)

I've added some comments I'd like to resolve first, and I've requested a review from @erik-wramner as weill, which I would like to see before merging.

**Comment by MakerMatrix** [comment]
On reviewing your comments I think it would be much better if we wrapped the the clamav alias inside an if stmt, to check whether the user asked for that feature or not.  I put it in because I noticed that cron job(s) get run as the clamav user.  But I guess only if user asked for that feature.

**Comment by aendeavor** [comment]
> On reviewing your comments I think it would be much better if we wrapped the the clamav alias inside an if stmt, to check whether the user asked for that feature or not. I put it in because I noticed that cron job(s) get run as the clamav user. But I guess only if user asked for that feature.

I accidentally did not press the submit button. You can now see my comments, sorry:) These are the ones I was referring to.

**Comment by aendeavor** [comment]
In #1706 you mentioned that `/etc/aliases` is completely emptied on "check-for-changes.sh`-runs. I've already mentioned it in a comment - this causes some headache. I cannot confidently say whether this may break things.

**Comment by MakerMatrix** [comment]
> In #1706 you mentioned that `/etc/aliases` is completely emptied on "check-for-changes.sh`-runs. I've already mentioned it in a comment - this causes some headache. I cannot confidently say whether this may break things.

Fair question, but if you look closely you see that line was already there - I only made the comment, and then added the two lines below to append amavis and clamav.  It was always the case that /etc/aliases got recreated from scratch when these functions are called.  There are some comments about this and a TODO about making these things atomic instead of rebuilding them from scratch, [here](https://github.com/tomav/docker-mailserver/blob/2629b575900759920a31ad6a5c0603cbc7b25af7/target/check-for-changes.sh#L56).

**Comment by aendeavor** [comment]
Oh my... I totally missed that. I didn't look close enough. Resolved:)

File: target/start-mailserver.sh [hunk 1]
```diff
@@ -1031,8 +1031,12 @@ s/$/ pcre:\/etc\/postfix\/regexp/
 
   _notify 'inf' "Configuring root alias"
 
+  # Generate default /etc/aliases
   echo "root: ${POSTMASTER_ADDRESS}" > /etc/aliases
```
[the remark this answers is not in the document]
Comment (aendeavor) [reply target/start-mailserver.sh:1035]: Turns out: I didn't look close enough :D

File: Dockerfile [hunk 2]
```diff
@@ -197,7 +200,7 @@ RUN mkdir /var/run/fetchmail && chown fetchmail /var/run/fetchmail
 # Configures Postfix
 COPY target/postfix/main.cf target/postfix/master.cf /etc/postfix/
 COPY target/postfix/header_checks.pcre target/postfix/sender_header_filter.pcre target/postfix/sender_login_maps.pcre /etc/postfix/maps/
-RUN echo "" > /etc/aliases
+RUN touch /etc/aliases
```
[the remark this answers is not in the document]
Comment (MakerMatrix) [reply Dockerfile:203]: It was not clear to me why that line is needed at all.  I made the assumption that the goal was to create the file anew, since it's the Dockerfile here.  I changed it because I thought the old way was clunky.

I would have liked to have the defaults put in here but I don' think the environment for $HOSTNAME and $DOMAIN are exposed inside the Dockerfile?  I had tested versions that did things like:

amavis: postmaster

...etc...but that syntax in /etc/aliases forced the mail to get delivered locally (/var/mail) or fail if there is no user (i.e. postmaster).  Only if you specify something after @, will it deliver through SMTP.

[the remark this answers is not in the document]
Comment (aendeavor) [reply Dockerfile:203]: I understand. I'd propose the following to get the original functionality without the clunkyness:

``` DOCKERFILE
RUN : >/etc/aliases
```

[the remark this answers is not in the document]
Comment (MakerMatrix) [reply Dockerfile:203]: I'm not a bash user, really.  If you say it does that, then I like its cleanliness.

How do I update it now that it's already a PR?  Just push this change to the same branch?

[the remark this answers is not in the document]
Comment (aendeavor) [reply Dockerfile:203]: > I'm not a bash user, really.  If you say it does that, then I like its cleanliness.

I just wrote a little test Dockerfile - my suggested solution should work.

> How do I update it now that it's already a PR?  Just push this change to the same branch?

Exactly:)

File: target/start-mailserver.sh [hunk 3]
```diff
@@ -122,6 +122,7 @@ function register_functions
 
   _register_setup_function "_setup_dkim"
   _register_setup_function "_setup_ssl"
+  _register_setup_function "_setup_postfix_aliases"
```
Comment (aendeavor) [inline target/start-mailserver.sh:125]: What's this about?:)

File: target/check-for-changes.sh [hunk 4]
```diff
@@ -79,8 +79,15 @@ do
         esac
       done
 
-      # regenerate postix aliases
+      # Generate default /etc/aliases
       echo "root: ${PM_ADDRESS}" >/etc/aliases
+      echo "amavis: ${PM_ADDRESS}" >>/etc/aliases
+      if [[ ${ENABLE_CLAMAV} -1 1 ]]
```
Comment (aendeavor) [inline target/check-for-changes.sh:85]: this is invalid, you accidentally used `-1` instead of `-eq` :)

Comment (MakerMatrix) [reply target/check-for-changes.sh:85]: OMG

File: target/start-mailserver.sh (hunk 3 above)
Comment (MakerMatrix) [reply target/start-mailserver.sh]: Sloppiness.  I was moving it around to test why aliases do not get created when using default mailserver.env.  It's still a mystery.

**Comment by aendeavor** [comment]
Looks somewhat fine to me now. As mentioned in #1706 I am not sure whether the change now adds the aliases for root, amavis and clam, but I really don't see why this would be case. So is this working?:)

**Comment by aendeavor** [comment]
What's the status here? @erik-wramner  what do you think?

File: Dockerfile [hunk 5]
```diff
@@ -197,7 +200,7 @@ RUN mkdir /var/run/fetchmail && chown fetchmail /var/run/fetchmail
 # Configures Postfix
 COPY target/postfix/main.cf target/postfix/master.cf /etc/postfix/
 COPY target/postfix/header_checks.pcre target/postfix/sender_header_filter.pcre target/postfix/sender_login_maps.pcre /etc/postfix/maps/
-RUN echo "" > /etc/aliases
+RUN : >/etc/aliases
```
Comment (casperklein) [inline Dockerfile:203]: Why not just `RUN >/etc/aliasses` which is even shorter ;-)

Comment (aendeavor) [reply Dockerfile:203]: True indeed, and that would work, but I think we talked about this a while ago :D This nooo `:` makes the intention clear and is idiomatic Bash :)

Comment (casperklein) [reply Dockerfile:203]: Now that you mention it, I can remember 😇  But _personally_ I don't like using functionless commands like in that case and don't see how that makes the intention any clearer. Using something like `while :; do foo; done` is a better noop use case.
But in the end, it's not worth discussing about +/- 2 bytes 😅 👍

Comment (aendeavor) [reply Dockerfile:203]: +1

**Comment by MakerMatrix** [comment]
> Looks somewhat fine to me now. As mentioned in #1706 I am not sure whether the change now adds the aliases for root, amavis and clam, but I really don't see why this would be case. So is this working?:)

Yeah it works, in conjuction with #1707. In 1706 I had made some faulty assumptions about the default testing container config, just learning how to build and test. :).  It was a simple matter of getting a a couple things configured away from defaults which in any real-world deployment would be the first things somebody would set.

**Comment by aendeavor** [comment]
@erik-wramner when you've reviewed this PR, feel free to either merge it or close it. SInce we are working on the alias code anyways, I will leave it up to you what to do with this PR.

File: Dockerfile [hunk 6]
```diff
@@ -134,6 +135,8 @@ RUN sed -i -e 's/include_try \/usr\/share\/dovecot\/protocols\.d/include_try \/e
   # create directory for certificates created by mkcert
   mkdir /etc/dovecot/ssl && \
   chmod 755 /etc/dovecot/ssl  && \
+  touch /etc/dovecot/ssl/.rand && \
+  sed -i 's/dev\/urandom/etc\/dovecot\/ssl\/\.rand/' /usr/share/dovecot/dovecot-openssl.cnf && \
```
Comment (erik-wramner) [inline Dockerfile:139]: Why are you changing the dovecot random function in this way and how is it related to /etc/aliases? If this changes should be done (and please explain why) it should probably go in a separate commit.

File: Dockerfile [hunk 7]
```diff
@@ -207,6 +210,9 @@ RUN sed -i -r "/^#?compress/c\compress\ncopytruncate" /etc/logrotate.conf && \
   chown -R clamav:root /var/log/mail/clamav.log && \
   touch /var/log/mail/freshclam.log && \
   chown -R clamav:root /var/log/mail/freshclam.log && \
+  mkdir -p /var/log/supervisor && \
+  touch /var/log/supervisor/supervisord.log && \
+  chown -R syslog:root /var/log/supervisor && \
```
Comment (erik-wramner) [inline Dockerfile:215]: Please explain the need for this as well and how it is related to the aliases. It looks like you are trying to solve something, but what?

File: target/check-for-changes.sh [hunk 8]
```diff
@@ -79,8 +79,15 @@ do
...
+      # Generate default /etc/aliases
       echo "root: ${PM_ADDRESS}" >/etc/aliases
+      echo "amavis: ${PM_ADDRESS}" >>/etc/aliases
+      if [[ ${ENABLE_CLAMAV} -eq 1 ]]
+      then
+        echo "clamav: ${PM_ADDRESS}" >>/etc/aliases
+      fi
+
+      # Append external postifx aliases
       if [[ -f /tmp/docker-mailserver/postfix-aliases.cf ]]
```
Comment (erik-wramner) [inline target/check-for-changes.sh:91]: This is probably fine, but we could check if postfix-aliases.cf has a mapping for amavis. Many existing installations may already have one (mine does). Not a big deal though, perhaps it is better to keep the code simple and write something about it in the README.

File: target/start-mailserver.sh [hunk 9]
```diff
@@ -1031,8 +1031,15 @@ s/$/ pcre:\/etc\/postfix\/regexp/
 
   _notify 'inf' "Configuring root alias"
 
+  # Generate default /etc/aliases
   echo "root: ${POSTMASTER_ADDRESS}" > /etc/aliases
+  echo "amavis: ${POSTMASTER_ADDRESS}" >> /etc/aliases
+  if [[ ${ENABLE_CLAMAV} -eq 1 ]]
+  then
+    echo "clamav: ${POSTMASTER_ADDRESS}" >> /etc/aliases
+  fi
```
Comment (erik-wramner) [inline target/start-mailserver.sh:1041]: See previous comment.

**Comment by erik-wramner** [comment]
@aendeavor @MakerMatrix I don't understand what several of the changes are for. Perhaps the important fixes are done elsewhere? If so let's close this. If not feel free to motivate and perhaps we can merge instead. It feels as if the changes are trying to fix several things that are not related to the alias issue at all?

**Comment by aendeavor** [comment]
Very true. There are some hotfixes in here that could probably go in another PR, you're right. @MakerMatrix can explain them and then we can decide if we need another PR.

File: Dockerfile (hunk 6 above)
Comment (MakerMatrix) [reply Dockerfile:139]: This is an unrelated build fix that I tossed in quickly.  As I recall, during the build phase, openssl throws an error when generating the dovecot key.  IIRC it's because it tries to write entropy to  /dev/urandom which of course is not allowed.  So now I have it creating an empty, writable file for it to put its entropy in, and that filename gets put in dovecot-openssl.cnf.

File: Dockerfile (hunk 7 above)
Comment (MakerMatrix) [reply Dockerfile:215]: It's not related to aliases, sorry.  I believe I found errors in the logs where supervisord was throwing an error when trying to write to /var/log/supervisor/supervisord.log and it did not exist.  I do apologize for not documenting this one better, I see I didn't even put a comment along with the PR and frankly even I cannot remember exactly where I found this one.

File: target/check-for-changes.sh (hunk 8 above)
Comment (MakerMatrix) [reply target/check-for-changes.sh:91]: Hmmm, mine didn't.   My /tmp/dockermailserver/postfix-aliases.cf is empty.  If that's a better place to do these, then by all means.

File: target/start-mailserver.sh (hunk 9 above)
Comment (MakerMatrix) [reply target/start-mailserver.sh:1041]: Ditto.

File: Dockerfile (hunk 7 above)
Comment (MakerMatrix) [reply Dockerfile:215]: I'll amend the first sentence of the previous to say that it *was* tangentially related to aliases in the sense that I found it when trying to discover exactly what was going on with aliases when supervisord brings up the mail service the first time after container creation.  There are some conditions where they don't get added, which appear to be based on how some of the environment variables are set.  I never narrowed it down further than that but thought it important for supervisord to be able to write its logs "out of the box"

File: Dockerfile (hunk 6 above)
Comment (casperklein) [reply Dockerfile:139]: That is the build message:

    Generating a RSA private key
    ....+++++
    ........+++++
    writing new private key to '/etc/dovecot/ssl/dovecot.key'
    -----
    Cannot write random bytes:
    140123681975557:error:2407007A:random number generator:RAND_write_file:Not a regular file:../crypto/rand/randfile.c:183:Filename=/dev/urandom

If I understand https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=922732#20 correctly, it should be safe to remove RANDFILE completely from the configuration:

    | OpenSSL 1.1.1 introduced a new random generator (CSPRNG) with an improved
    | seeding mechanism. The new seeding mechanism makes it unnecessary to
    | define a RANDFILE for saving and restoring randomness. This option is
    | retained mainly for compatibility reasons.

In any case, this should be discussed/done in a separate PR.

Comment (aendeavor) [reply Dockerfile:139]: @casperklein I'm completely with you there.

@MakerMatrix when we have finished the migration to https://github.com/docker-mailserver/docker-mailserver, could open a few several small PRs that go through each of your changes one by one?

Comment (aendeavor) [reply Dockerfile:139]: @casperklein @MakerMatrix I'm currently working on a bigger improvement of the `Dockerfile`. @casperklein How would I go about removing the RANDFILE completely?

Comment (aendeavor) [reply Dockerfile:139]: @casperklein Nevermind - I did it by removing `RANDFILE` from `mkcert.sh`. Done.

Comment (MakerMatrix) [reply Dockerfile:139]: > @MakerMatrix when we have finished the migration to https://github.com/docker-mailserver/docker-mailserver, could open a few several small PRs that go through each of your changes one by one?

Happy to.

**Comment by MakerMatrix** [comment]
At this point I think this can be closed out without merging.  I will try to recreate the supervisord.log issue and open that as a separate PR after the migration to the docker-mailserver github account (will I know when that happens?).  I think everything else here including the additional default aliases is being addressed in different fixes.

**Comment by aendeavor** [comment]
> At this point I think this can be closed out without merging.  I will try to recreate the supervisord.log issue and open that as a separate PR after the migration to the docker-mailserver github account (will I know when that happens?).  I think everything else here including the additional default aliases is being addressed in different fixes.

A good decision. You will see when the repository migration is ready in the README and when the other PRs are all closed - just look out for that.
